File: Algorithms/0583_Delete_Operation_for_Two_Strings/Python/Delete_Operation_for_Two_Strings_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def minDistance(self, word1, word2):
        length1, length2 = len(word1), len(word2)
        if length1 == length2 == 0: return 0
        if word1 == word2: return 0

        board = [[0 for _ in range(length1 + 1)] for _ in range(length2 + 1)]
        for y in range(length2 + 1):
            for x in range(length1 + 1):
                if x == 0 or y == 0: continue
                if word1[x - 1] == word2[y - 1]:
                    board[y][x] = board[y - 1][x - 1] + 1
                else:
                    board[y][x] = max(board[y - 1][x], board[y][x - 1])
        res = board[-1][-1]

        return length1 - res + length2 - res

File: Algorithms/0583_Delete_Operation_for_Two_Strings/Java/DeleteOperationForTwoStringsSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int minDistance(String word1, String word2) {
        if (word1.equals(word2)) return 0;
        int[][] board = new int[word1.length() + 1][word2.length() + 1];
        for (int[] ints : board) {
            Arrays.fill(ints, 0);
        }
        for (int y = 1; y <= word1.length(); y++) {
            for (int x = 1; x <= word2.length(); x++) {
                if (word1.charAt(y - 1) == word2.charAt(x - 1)) {
                    board[y][x] = board[y - 1][x - 1] + 1;
                } else {
                    board[y][x] = Math.max(board[y][x - 1], board[y - 1][x]);
                }
            }
        }
        return word1.length() + word2.length() - 2 * board[word1.length()][word2.length()];
    }
}




